File: 0077-combinations/0077-combinations.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void helper(int index, int n, int k, List<Integer> curr) {
        if (k == 0) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        if (n - index + 1 < k) return;
        for (int i = index; i <= n; i++) {
            curr.add(i);
            helper(i+1, n, k-1, curr);
            curr.remove(curr.size() - 1);
        }
    }
    public List<List<Integer>> combine(int n, int k) {
        helper(1, n, k, new ArrayList<>());
        return ans;
    }
}

File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    vector<vector<int>> ans;

    void dfs(int x, int n, int k, vector<int>& path) {
        if (k == 0) { 
            ans.push_back(path); 
            return; 
        }
        if (n - x + 1 < k) return;
        for (int i = x; i <= n; i++) {
            path.push_back(i);
            dfs(i+1, n, k-1, path);
            path.pop_back();
        }
    }

    vector<vector<int>> combine(int n, int k) {
        vector<int> path;
        dfs(1, n, k, path);
        return ans;
    }
};
